src/app/clusters/thread-network-diagnostics-server/thread-network-diagnostics-provider.cpp

```diff
@@ -0,0 +1,861 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "thread-network-diagnostics-provider.h"
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/data-model/Encode.h>
+#include <lib/support/CodeUtils.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#if (CHIP_DEVICE_CONFIG_ENABLE_THREAD && !CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK)
+#include <platform/OpenThread/OpenThreadUtils.h>
+#include <platform/ThreadStackManager.h>
+
+#if CHIP_DEVICE_CONFIG_THREAD_FTD
+#include <openthread/thread_ftd.h>
+#endif // CHIP_DEVICE_CONFIG_THREAD_FTD
+#endif // (CHIP_DEVICE_CONFIG_ENABLE_THREAD && !CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK)
+
+using namespace chip::DeviceLayer;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ThreadNetworkDiagnostics {
+
+/*
+ * @brief Get runtime value from the thread network based on the given attribute ID.
+ *        The info is encoded via the AttributeValueEncoder.
+ *
+ * @param attributeId Id of the attribute for the requested info.
+ * @param aEncoder Encoder to encode the attribute value.
+ *
+ * @return CHIP_NO_ERROR = Succes.
+ *         CHIP_ERROR_NOT_IMPLEMENTED = Runtime value for this attribute to yet available to send as reply
+ *                                      Use standard read.
+ *         CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE = Is not a Runtime readable attribute. Use standard read
+ *         All other errors should be treated as a read error and reported as such.
+ *
+ * @note This function implementation can compile in 3 different outcome
+ *       (1) CHIP_DEVICE_CONFIG_ENABLE_THREAD && !CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK:
+ *           - Generic implementation fetching the valid thread network data from the thread stack and encoding it respectively to
+ *             the attributeID received.
+ *       (2) CHIP_DEVICE_CONFIG_ENABLE_THREAD && CHIP_DEVICE_CONFIG_USES_OTBR_POSIX_DBUS_STACK:
+ *           - Encode a NULL value for nullable attributes or 0 for the others.
+ *           - Devices using the ot-br-posix dbus stack have not yet provided the API the fetch the needed thread informations.
+ *       (3) None of the conditions above
+ *           - returns CHIP_ERROR_NOT_IMPLEMENTED.
+ */
+CHIP_ERROR WriteThreadNetworkDiagnosticAttributeToTlv(AttributeId attributeId, app::AttributeValueEncoder & encoder)
```

Comment:
This function could have been simplified by having three sub functions that can compiled in or out based on the define values. having all 3 outcomes in one function makes it pretty huge.

Reply:
We wouldn't want to compile the 2 of the 3 extra functions so splitting it in 3 would be just adding more #ifdef to the code. 

Two of those functions don't have much logic. Return error. and the other encode the correct default value. 

Not convinced splitting it is a better option